packages/cozy-stack-client/src/Collection.js

```diff
@@ -43,7 +43,7 @@ export class Collection {
     try {
       const resp = await stackClient.fetchJSON(method, endpoint)
       return {
-        data: normalize(resp.data)
+        data: normalize(resp.data || resp)
```

Comment:
Je trouve ça très magique et je n'aime pas trop. On fait quoi si on est dans un des cas sans JSON API mais où le document couch a une sous-clef `data` ?

Reply:
Comment tu reconnais les cas sans JSON API ? Via une whitelist ?

Reply:
Je ne sais pas justement. 

Je vois plusieurs possibilités : 

1. On normalise `resp` dans tous les cas et c'est éventuellement à l'appelant de savoir qu'il fait du JSON API et utiliser la sous-clef `data`
2. On va faire de l'auto-magique pour garder compatibilité avec l'existant et on ajoute un paramètre `useJsonApi` (nom à trouver) qui est à `true` par défaut. Charge à l'appelant de savoir qu'il ne fait pas de JSON API et d'ajouter cette option à `false`
3. On fait de `get` une méthode d'instance (éventuellement sous un autre nom), qui va regarder dans une propriété `useJsonApi` (nom à trouver) de l'objet source pour savoir quoi faire

Du point de vue architecture je pense qu'on aurait du utiliser la (3). Là on utilise des méthodes statiques alors que la relation qui semble la bonne est à priori un héritage (DocumentCollection > Collection).

Sur une architecture en méthodes utilitaires statiques, j'aurais utilisée la (1) mais ça me parait casser inutilement de l'historique. La (2) me suffirait à priori.

Reply:
Oh, et en fait j'ai un (4) qui me parait peut-être aussi bien : 

`normalize(data, resp)` 

Charge à l'appelant de savoir quoi utiliser, et ceux qui n'utilisent que `data` n'ont rien à changer

Reply:
Je suis d'accord avec @edas 

Personnellement, je trouve que cette abstraction de Collection n'apporte pas grand chose si ce n'est beaucoup de magie. Et j'aime pas trop ça. 

Je préférai la solution précédente où chaque collection gérait elle même son get et son throw en fonction de ce qu'elle sait (et elle sait plus sur la structure de la réponse que cette méthode abstraite de Collection)

Reply:
Le but de cette abstraction était de regrouper beaucoup de morceaux de code identiques issus de copié-collé-adapté. Je n'ai pas de volonté forte à vouloir la garder.

Merci @edas pour tes propositions, j'ai failli écrire que je n'étais pas satisfait de cette solution, que j'ai implémenté un peu vite pour proposer un fix.

On part donc sur la 4) ?

Reply:
Pour moi la 4 est la mieux effectivement.

P/r à l'architecture (mais c'est un autre problème), j'ai l'impression que : 

- `get` devrait revenir dans DocumentCollection
- Collection devrait disparaître
- `get` devrait utiliser `this.normalize`
- `normalize` devrait être une méthode d'instance

Reply:
@edas J'ai poussé un nouveau commit. J'ai toujours un ou logique qui est présent, cette fois au sein de `normalize``, et donc un peu moins grossier. Est-ce que tu vois une façon d'encore améliorer cette implémentation ?

Reply:
Le but ce n'est pas d'avoir une AlbumCollection ou le normalize prend `resp` et non `resp.data` ?

Reply:
Pour moi tu ne fais que reporter le problème à la couche d'au dessus dans ton dernier commit.

Tel que je le vois tu devrais avoir : 
- Un DocumentCollection sans la fonction de normalisation
- Un JsonApiDocumentCollection, dérivé de DocumentCollection, avec la fonction de normalisation qui utilise resp.data
- Un RawApiDocumentCollection, dérivé de DocumentCollection, avec la fonction de normalisation qui utilise resp directement

Ensuite chaque collection étend soit JsonApiDocumentCollection soit RawApiDocumentCollection.

Si vous voulez aller plus vite on peut considérer que DocumentCollection implémente à la base une des deux solutions, et ne gérer que le cas particulier via une classe dérivée.

Dans tous les cas il faut que quand tu utilises une collection, ta collection sache si elle utilise directement les réponses ou si elle passe par une sous-entrée data. Tu ne devrais pas avoir à gérer un "je teste et si ce n'est pas ça j'utilise autre chose"

Reply:
Donc finalement, plus vraiment la 4). En relisant le code et la documentation de la stack sur le data system https://github.com/cozy/cozy-stack/blob/master/docs/data-system.md#request, il semble qu'en effet la solution basée sur l'héritage de deux collections `JsonApiDocumentationCollection` et `RawApiDocumentationCollection` soit complètement appropriée. Cela aurait dû être ma réflexion dans ce commit https://github.com/cozy/cozy-client/commit/8ae9740a06a6246593027fc0eae89c804ce88a99#diff-895069b14d04da2025d5c477287dd819.

Je regarde les implications plus en profondeur et vous fait un rapide retour ici.